src/stdio/src/lib.rs

```diff
@@ -1,41 +1,64 @@
 //! stdio implementation for Redox, following http://pubs.opengroup.org/onlinepubs/7908799/xsh/stdio.h.html
 
+#![feature(compiler_builtins_lib)]
 #![no_std]
 
+extern crate compiler_builtins;
 extern crate errno;
+extern crate fcntl;
 extern crate platform;
+extern crate stdlib;
+extern crate string;
 extern crate va_list as vl;
 
 use core::str;
+use core::ptr;
 use core::fmt::Write;
+use core::sync::atomic::{AtomicBool, Ordering};
 
 use platform::types::*;
-use platform::c_str;
-use platform::errno;
+use platform::{c_str, errno};
 use errno::STR_ERROR;
 use vl::VaList as va_list;
 
 mod printf;
 
-pub const BUFSIZ: c_int = 4096;
+mod default;
+pub use default::*;
 
-pub const FILENAME_MAX: c_int = 4096;
+mod constants;
+pub use constants::*;
 
-pub type fpos_t = off_t;
+mod helpers;
 
-pub struct FILE;
+mod internal;
 
-#[allow(non_upper_case_globals)]
-#[no_mangle]
-pub static mut stdout: *mut FILE = 1 as *mut FILE;
-
-#[allow(non_upper_case_globals)]
-#[no_mangle]
-pub static mut stderr: *mut FILE = 2 as *mut FILE;
+/// The structure of this struct is likely to change.
+/// all of the _pos pointers are likely to change to isize offsets and its possible that the
+/// function pointers will be removed
+#[repr(C)]
+pub struct FILE {
+    flags: c_int,
+    rpos: *mut u8,
+    rend: *mut u8,
+    wend: *mut u8,
+    wpos: *mut u8,
+    wbase: *mut u8,
+    fd: c_int,
+    buf: *mut u8,
+    buf_size: size_t,
+    buf_char: i8,
+    lock: AtomicBool,
+    unget: size_t,
+    write: Option<*const (Fn(*mut FILE, *const u8, usize) -> size_t)>,
+    read: Option<*const (Fn(*mut FILE, *mut u8, usize) -> size_t)>,
+    seek: Option<*const (Fn(*mut FILE, off_t, c_int) -> off_t)>,
```

Comment:
If you change `write`, `read`, and `seek` to an `Option<fn(*mut FILE, off_t, c_int)>` I think it could make the code a lot more readable. If you do this, I think `cbindgen` will try to export the `FILE` structure, so you might have to add something to the [cbindgen.toml](https://github.com/eqrion/cbindgen#configuration) to ensure that it doesn't get exported, and then manually add a `typedef` to `bits/stdio.h`.

Reply:
I can't remember why I didn't have them as `Option<fn(...)>`. I think there was a reason at some point in time (possibly because I tried `*const fn(...)` by accident), but yes, this would be a great idea for readability!

Cheers!

Reply:
I have been thinking of just removing the function pointers though, as we currently only have one method of doing them anyway.